devito/finite_differences/derivative.py

```diff
@@ -497,3 +580,141 @@ def _eval_fd(self, expr, **kwargs):
+        # function supports.
+        for (d, do), (_, fo) in zip(dcounter.items(), fcounter.items(), strict=True):
+            if getattr(d, 'is_Time', False):
+                dim_name = 'time'
+                order = self.expr.time_order
+            else:
+                dim_name = 'space'
+                order = self.expr.space_order
+            # The `fd_order` may need to be reduced to construct the nested derivative
+            # in this case we only emit a warning
+            if fo > order:
+                if do > order:
+                    raise ValueError(
+                        f'Nested {do}-derivative constructed which is bigger '
+                        f'than the {dim_name}_order={order}'
+                    )
+                else:
+                    warn(
+                        f'Nested derivative constructed with fd_order={fo}, '
+                        f'but {dim_name}_order={order}. Adjusting '
+                        f'fd_order={order} for the {d} dimension.'
+                    )
+                    fcounter[d] = order
+
+        # 4. Finally, construct the new derivative object with the updated counts
+        # and kwargs.
+        new_kwargs = {
+            'deriv_order': tuple(dcounter.values()),
+            'fd_order': tuple(fcounter.values())
+        }
+        return self.func(new_expr, *dcounter.items(), **new_kwargs)
+
+    def _eval_expand_mul(self, **hints):
+        """
+        Expands products, moving independent terms outside the derivative
+        `Derivative(C·f(x)·g(c, y), x)
+            --> C·g(y)·Derivative(f(x), x)`
+        """
+        if self.expr.is_Mul:
+            ind, dep = self.expr.as_independent(*self.dims, as_Add=False)
+            return ind*self.func(dep)
+        else:
+            return self
+
+    def _eval_expand_add(self, **hints):
+        """
+        Expands sums, using linearity of derivative
+        `Derivative(f(x) + g(x), x)
+            --> Derivative(f(x), x) + Derivative(g(x), x)`
+        """
+        if self.expr.is_Add:
+            ind, dep = self.expr.as_independent(*self.dims, as_Add=True)
+            if dep.is_Add:
+                return Add(*[self.func(s, *self.args[1:]) for s in dep.args])
+            else:
+                return self.func(dep)
+        else:
+            return self
+
+    def _eval_expand_product_rule(self, **hints):
+        """
+        Expands products, of functions of the dependent variable
+        `Derivative(f(x)·g(x), x)
+            --> Derivative(f(x), x)·g(x) + f(x)·Derivative(g(x), x)`
+        This is only implemented for first derivatives with an arbitrary number
+        of multiplicands and second derivatives with two multiplicands. The
+        resultant expression for higher derivatives and mixed derivatives is much
+        more difficult to implement.
+        """
+        if self.expr.is_Mul and len(self.dims) == 1:
+            if self.deriv_order == (1,):
```

Comment:
You can do `args=self.expr.args` before those to make it less verbose and avoid the `self.expr.args` everywhere